### DISCUSSION TITLE
Authorizing users via GitHub app, missing user emails.

### DISCUSSION THREAD

**Discussion by gylaz** [community/community#24843] [Apps, API and Webhooks]
<p>We’ve recently switched from an Oauth app to GitHub app for everything, including authenticating in.</p>
<p>We noticed that most of the time the user email comes in as nil/undefined. Not all, however, but majority.</p>
<p>Any one have an idea of why we could be missing user emails? Do we need a certain setting enabled in our GitHub app configuration? Or do emails not get exposed when user authenticates with our GitHub app? If so, what can we do to regain access to emails?</p>

**Comment by webknjaz** [comment]
<p>There’s “Emails” entry in the GitHub App settings under “Permissions &amp; events” -&gt; “User permissions”. I think, you should set it to read-only.</p>

**Comment by gylaz** [comment]
<p>Thanks for the reply! I believe the “Emails” permission setting is for accessing user emails via the API call.</p>
<p>To clarify, I’m talking about when the user first authorizes/signs-in via the GitHub app, there’s information about the user we get from GitHub (in our case, <code>omniauth.auth</code> in the Rails request hash). What’s odd, is we still get emails for some users after the login, but for most emails are missing. This seems to have started happening after we switched from Oauth app to GitHub app (early June, if some additional GitHub feature was released at that time).</p>
<p>I also noticed that GitHub now has a <strong>Keep my email address private</strong> setting in <a href="https://github.com/settings/emails" rel="nofollow noopener">https://github.com/settings/emails</a> but it seems to be unchecked (by default?), unless the user takes action. I suspect that isn’t the cause, as it’s unlikely most of our users enabled that option suddenly.</p>

**Comment by gylaz** [comment, marked as answer]
<p>If anyone comes across this, I got some help from GitHub via another channel. Here’s the gist:</p>
<p>The <a href="https://github.com/omniauth/omniauth" rel="nofollow noopener">omniauth-github</a> gem that I’m using is depending on OAuth scopes to determine if it should ask <a href="https://github.com/omniauth/omniauth-github/blob/master/lib/omniauth/strategies/github.rb#L66-L71" rel="nofollow noopener">for user emails</a>:</p>
<pre><code>def email_access_allowed?
return false unless options['scope']
email_scopes = ['user', 'user:email']
scopes = options['scope'].split(',')
(scopes &amp; email_scopes).any?
end
</code></pre>
<p>If it doesn’t find the oauth scope it relies on the email found in <a href="https://github.com/omniauth/omniauth-github/blob/master/lib/omniauth/strategies/github.rb#L45-L48" rel="nofollow noopener">the user profile</a> which is <a href="https://github.com/settings/profile" rel="nofollow noopener">a different user setting</a> that defaults to not selected. A workaround to this would be to configure omniauth to request <code>user:emailscope</code>. It will be ignored by GitHub as GitHub Apps don’t use scopes but it would trigger <code>omniauth</code> to hit the user/emails API. Still need to request the user emails permission mentioned in this forum post.</p>

**Reply by jclusso** [reply]
The scope is `user:email` not `user:emailscope`.
